File: 程式設計( 二 )/王國華/期末考練習/1. MagicSquareClass/MagicSquare.cpp
#include <iostream>
#include <iomanip>

using namespace std;

class Magicsquare{
    public:
        Magicsquare(int n) : size(n), row(0), col( (n + 1) / 2 ) {
            for ( int i = 0; i < size; i++ ){
                array[i] = new int[size];
            }
        }
        ~Magicsquare(){
            for ( int i = 0; i < size; i++ ){
                delete[] array[i];
            }
            delete[] array;
        };

        void fillsquare(){
            for ( int key = 1; key <= size * size; key++ ){
                if ( key % size == 1 ){
                    row++;
                }else{
                    row--;
                    col++;
                }

                if ( row == 0 ){
                    row = size;
                }
                if ( col > size ){
                    col = 1;
                }
                array[row - 1][col - 1] = key;
            }
        }

        void printsquare(){
            cout << (1 + size * size) * size / 2 << endl;
            for ( int i = size - 1; i >= 0; i-- ){
                for ( int j = 0; j < size; j++ ){
                    if ( j != size - 1 ){
                        cout << array[i][j] << " " ;
                    }else{
                        cout << array[i][j];
                    }    
                }
                if ( i != size - 1 ){
                    cout << endl;
                }
            }
        }
    private:
        const int size;
        int row, col;
        int **array = new int*[size];
};

int main(){
    int n;
    int flag = 0;
    while ( cin >> n ){
        Magicsquare square(n);
        
        if ( flag != 0 ){
            cout << endl;
        }

        if ( n % 2 == 0 ){
            cout << "It is not an odd number." << endl;
            flag = 1;
        }else{
            square.fillsquare();
            square.printsquare();
            flag = 1;
        }
    }
}

File: 程式設計( 三 )/6. Magic Square/Magic_Square.java
import java.util.Scanner;

class Square {
    private int size;
    private int row;
    private int col;
    private int[][] arr;

    public Square( int n ){
        size = n;
        row = n - 1;
        col = n / 2;
        arr = new int[size][size];
    }


    public void fillsquare(){
        for ( int key = 1; key <= size * size; key++ ){
            if ( row < 0 ){
                row = size - 1;
            }
            if ( col < 0 ){
                col = size - 1;
            }
            if ( row >= size ){
                row = 0;
            }
            if ( col >= size ){
                col = 0;
            }

            arr[row][col] = key;

            if ( key % size == 0 ){
                row--;
            }else{
                row++;
                col++;
            }
        }
    }

    public void printsquare(){
        System.out.println( (1 + size * size ) * size / 2 );
        for ( int i = 0; i < size; i++ ){
            for ( int j = 0; j < size; j++ ){
                System.out.printf("%5d", arr[i][j]);
            }
            System.out.println();
        }
    }
    
}


public class Magic_Square {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        int flag = 0;
        while ( input.hasNext() ){
            int n = input.nextInt();
            if ( flag == 0 ){
                flag = 1;
            }else{
                System.out.println();
            }

            if ( n % 2 == 0 ){
                System.out.println("It is not an odd number.");
            }else{
                Square square = new Square(n);
                square.fillsquare();
                square.printsquare();
            }
        }
        input.close();
    }
}